File: macos/HelixBrowser/UrlUtils.swift
import Foundation

struct UrlUtils {
    static func isUrl(_ input: String) -> Bool {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.hasPrefix("http://") ||
               trimmed.hasPrefix("https://") ||
               trimmed.hasPrefix("ftp://") ||
               (trimmed.contains(".") && !trimmed.contains(" "))
    }
    
    static func formatUrl(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return "helix://start" }
        if trimmed.hasPrefix("helix://") { return trimmed }
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return trimmed
        }
        if trimmed.hasPrefix("file://") || trimmed.hasPrefix("about:") {
            return trimmed
        }
        if isUrl(trimmed) {
            return "https://\(trimmed)"
        }
        return buildSearchQuery(trimmed, engine: Prefs.shared.searchEngine)
    }
    
    static func buildSearchQuery(_ query: String, engine: String) -> String {
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        switch engine.lowercased() {
        case "bing":
            return "https://www.bing.com/search?q=\(encoded)"
        case "duckduckgo":
            return "https://duckduckgo.com/?q=\(encoded)"
        case "yahoo":
            return "https://search.yahoo.com/search?p=\(encoded)"
        case "brave":
            return "https://search.brave.com/search?q=\(encoded)"
        default:
            return "https://www.google.com/search?q=\(encoded)"
        }
    }
    
    static func getDisplayUrl(_ url: String) -> String {
        guard let host = URL(string: url)?.host else { return url }
        return host.hasPrefix("www.") ? String(host.dropFirst(4)) : host
    }
    
    static func getFaviconUrl(_ url: String) -> String {
        guard let host = URL(string: url)?.host else { return "" }
        return "https://www.google.com/s2/favicons?domain=\(host)&sz=64"
    }
    
    static func isSearchQuery(_ input: String) -> Bool {
        return !isUrl(input)
    }
}

File: linux/src/url_utils.py
"""URL utility functions for Helix Browser"""
from urllib.parse import quote_plus, urlparse


class UrlUtils:
    @staticmethod
    def is_url(input_str: str) -> bool:
        trimmed = input_str.strip()
        return (trimmed.startswith("http://") or trimmed.startswith("https://") or
                trimmed.startswith("ftp://") or ("." in trimmed and " " not in trimmed))

    @staticmethod
    def format_url(input_str: str, engine: str = "google") -> str:
        trimmed = input_str.strip()
        if not trimmed:
            return "helix://start"
        if trimmed.startswith("helix://"):
            return trimmed
        if trimmed.startswith(("http://", "https://", "file://", "about:")):
            return trimmed
        if UrlUtils.is_url(trimmed):
            return f"https://{trimmed}"
        return UrlUtils.build_search_query(trimmed, engine)

    @staticmethod
    def build_search_query(query: str, engine: str) -> str:
        encoded = quote_plus(query)
        engines = {
            "bing": f"https://www.bing.com/search?q={encoded}",
            "duckduckgo": f"https://duckduckgo.com/?q={encoded}",
            "yahoo": f"https://search.yahoo.com/search?p={encoded}",
            "brave": f"https://search.brave.com/search?q={encoded}",
        }
        return engines.get(engine.lower(), f"https://www.google.com/search?q={encoded}")

    @staticmethod
    def get_display_url(url: str) -> str:
        try:
            parsed = urlparse(url)
            host = parsed.hostname or url
            return host[4:] if host.startswith("www.") else host
        except Exception:
            return url

    @staticmethod
    def get_favicon_url(url: str) -> str:
        try:
            host = urlparse(url).hostname
            return f"https://www.google.com/s2/favicons?domain={host}&sz=64" if host else ""
        except Exception:
            return ""
